wgpu-core/src/indirect_validation.rs

```diff
@@ -0,0 +1,369 @@
+use thiserror::Error;
+
+use crate::{
+    device::DeviceError,
+    pipeline::{CreateComputePipelineError, CreateShaderModuleError},
+};
+
+#[derive(Clone, Debug, Error)]
+#[non_exhaustive]
+pub enum CreateDispatchIndirectValidationPipelineError {
+    #[error(transparent)]
+    DeviceError(#[from] DeviceError),
+    #[error(transparent)]
+    ShaderModule(#[from] CreateShaderModuleError),
+    #[error(transparent)]
+    ComputePipeline(#[from] CreateComputePipelineError),
+}
+
+/// This machinery requires the following limits:
+///
+/// - max_bind_groups: 2,
+/// - max_dynamic_storage_buffers_per_pipeline_layout: 1,
+/// - max_storage_buffers_per_shader_stage: 2,
+/// - max_storage_buffer_binding_size: 3 * min_storage_buffer_offset_alignment,
+/// - max_push_constant_size: 4,
+/// - max_compute_invocations_per_workgroup 1
+///
+/// Which are be all indirectly satisfied by `DownlevelFlags::INDIRECT_EXECUTION`.
+#[derive(Debug)]
+pub struct IndirectValidation {
+    module: Box<dyn hal::DynShaderModule>,
+    dst_bind_group_layout: Box<dyn hal::DynBindGroupLayout>,
+    src_bind_group_layout: Box<dyn hal::DynBindGroupLayout>,
+    pipeline_layout: Box<dyn hal::DynPipelineLayout>,
+    pipeline: Box<dyn hal::DynComputePipeline>,
+    dst_buffer: Box<dyn hal::DynBuffer>,
+    dst_bind_group: Box<dyn hal::DynBindGroup>,
+}
+
+pub struct Params<'a> {
+    pub pipeline_layout: &'a dyn hal::DynPipelineLayout,
+    pub pipeline: &'a dyn hal::DynComputePipeline,
+    pub dst_buffer: &'a dyn hal::DynBuffer,
+    pub dst_bind_group: &'a dyn hal::DynBindGroup,
+    pub aligned_offset: u64,
+    pub offset_remainder: u64,
+}
+
+impl IndirectValidation {
+    pub fn new(
+        device: &dyn hal::DynDevice,
+        limits: &wgt::Limits,
+    ) -> Result<Self, CreateDispatchIndirectValidationPipelineError> {
+        let max_compute_workgroups_per_dimension = limits.max_compute_workgroups_per_dimension;
+
+        let src = format!("
+            @group(0) @binding(0)
+            var<storage, read_write> dst: array<u32, 3>;
+            @group(1) @binding(0)
+            var<storage, read> src: array<u32>;
+            struct OffsetPc {{
+                inner: u32,
+            }}
+            var<push_constant> offset: OffsetPc;
+
+            @compute @workgroup_size(1)
+            fn main() {{
+                let src = vec3(src[offset.inner], src[offset.inner + 1], src[offset.inner + 2]);
+                let res = select(src, vec3<u32>(), src > vec3({max_compute_workgroups_per_dimension}u));
+                dst[0] = res.x;
+                dst[1] = res.y;
+                dst[2] = res.z;
+            }}
+        ");
+
+        let module = naga::front::wgsl::parse_str(&src).map_err(|inner| {
+            CreateShaderModuleError::Parsing(naga::error::ShaderError {
+                source: src.clone(),
+                label: None,
+                inner: Box::new(inner),
+            })
+        })?;
```

Comment:
thought: I wonder if (as a lofty, low-priority optimization) we could move the parsing step here to build time (maybe prepared by `build.rs`?). I don't have a reason to suspect that this module compilation is problematic, and certainly won't block this PR on account of that. I guess I just wanted to throw that out there.

Reply:
It would be nice, it would also remove the need for the new `indirect-validation` feature flag. It's something I've never done before and wasn't sure how it would turn out. So, I can't assess it's worth right now.

Reply:
Ah, this was already discussed, sorry for missing it: https://github.com/gfx-rs/wgpu/pull/5714#issuecomment-2120047557